File: app/src/main/java/com/shxhzhxx/app/MainActivity.java
package com.shxhzhxx.app;

import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function3;

import android.util.Log;
import android.view.View;
import android.widget.TextView;

import com.shxhzhxx.urlloader.UrlLoader;
import com.shxhzhxx.urlloader.UrlLoaderEx;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Locale;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String TAG = "MainActivity";
    private static final String URL = "https://image.yizhujiao.com/FiZr1lFxhobKLogy4pkTfLqv6xrV";
    private static final String URL_BIG = "http://ovfjn9jer.bkt.clouddn.com/%E3%80%90%E4%B8%AD%E6%96%87%E5%85%AB%E7%BA%A7%E3%80%91%E4%B8%A4%E4%B8%AA%E5%9B%BD%E4%BA%BA%E5%B1%95%E5%BC%80%E4%BA%86%E6%83%8A%E4%BA%BA%E7%9A%84%E8%8B%B1%E8%AF%AD%E5%85%AB%E7%BA%A7%E5%AF%B9%E8%AF%9D_%28Av16874218_P1%29.mp4";

    private TextView mState;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        DownloadManager.init(this);

        mState = findViewById(R.id.state);
        findViewById(R.id.download).setOnClickListener(this);
        findViewById(R.id.cancel).setOnClickListener(this);
        findViewById(R.id.clear).setOnClickListener(this);

        UrlLoaderEx loader = new UrlLoaderEx(getCacheDir(),10*1024,4);
        loader.load("", "", file -> {
            Log.d(TAG,"onComplete");
            return null;
        }, () -> {
            Log.d(TAG,"onFailed");
            return null;
        }, () -> {
            Log.d(TAG,"onCanceled");
            return null;
        }, (aLong, aLong2, aLong3) -> {
            Log.d(TAG,"onProgressUpdate");
            return null;
        });
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.download:
                File file=new File(getCacheDir(),"shxhzhxx");
                try {
                    file.createNewFile();
                    file.setLastModified(123456);
                    Log.d(TAG, "lastModified: "+file.lastModified());
                    OutputStream os=new FileOutputStream(file);
                    os.write("fdsafdsa".getBytes());
                    os.close();
                    Log.d(TAG, "lastModified: "+file.lastModified());
                    file.setLastModified(123456);
                    Log.d(TAG, "lastModified: "+file.lastModified());
                    InputStream is=new FileInputStream(file);
                    is.read();
                    is.close();
                    Log.d(TAG, "lastModified: "+file.lastModified());

                } catch (IOException e) {
                    e.printStackTrace();
                }
                DownloadManager.getInstance().load(URL_BIG, new UrlLoader.ProgressObserver() {
                    @Override
                    public void onComplete(File file) {
                        mState.setText("onComplete: " + file.getAbsolutePath());
                    }

                    @Override
                    public void onFailed() {
                        mState.setText("onFailed");
                    }

                    @Override
                    public void onCanceled() {
                        mState.setText("onCanceled");
                    }

                    @Override
                    public void onProgressUpdate(long total, long current, long speed) {
                        mState.setText(String.format(Locale.CHINA, "onProgressUpdate: %.2f%%   %dKB/s", ((double) 100 * current / total), (int) (speed / 1024)));
                    }
                });
                break;
            case R.id.cancel:
                DownloadManager.getInstance().cancel(URL_BIG);
                Log.d(TAG, "onClick: cancel");
                break;
            case R.id.clear:
                DownloadManager.getInstance().clearCache(URL_BIG);
                break;
        }

    }
}

File: app/src/main/java/com/shxhzhxx/app/MainActivity.kt
package com.shxhzhxx.app

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.BitmapRegionDecoder
import android.graphics.Rect
import android.os.Bundle
import android.util.Log
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import com.shxhzhxx.urlloader.UrlLoader
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import java.io.File
import java.io.IOException
import java.util.concurrent.Executors
import java.util.concurrent.Future
import kotlin.coroutines.CoroutineContext

private const val TAG = "MainActivity"
private const val URL_BIG = "https://static.usasishu.com/bigFile.pdf"
private const val URL_BIG_VIDEO = "https://static.usasishu.com/bigVideoFile.mp4"
private const val URL_MAX_AGE = "http://plpwobkse.bkt.clouddn.com/ic_launcher.png"
private const val URL_BIG_IMG = "http://plpwobkse.bkt.clouddn.com/1125-2436-72.png"
private const val URL_IMG = "https://static.usasishu.com/image/2018/09/30/bg-index.jpg"
private const val URL_IMG_TINY = "https://static.usasishu.com/image/2018/09/29/img-grew-freely.png"

class MainActivity : AppCompatActivity(), CoroutineScope {
    private val job = SupervisorJob()
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy")
        job.cancel()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val loader = UrlLoader(cacheDir)
        val threadPool = Executors.newCachedThreadPool()
        asyncLoad.setOnClickListener {
            loader.asyncLoad(URL_IMG, onComplete = {
                Log.d(TAG, "asyncLoad onComplete:${it.absolutePath}")
                iv.scaleType = ImageView.ScaleType.FIT_CENTER
                iv.setImageBitmap(it.decodeBitmap(Params(it.absolutePath, iv.width, iv.height, true)))
            }, onProgress = { total, current, speed ->
                Log.d(TAG, "asyncLoad onProgress: total:$total    current:$current    speed:$speed")
            }, onCanceled = {
                Log.d(TAG, "asyncLoad onCanceled")
            }, onFailed = {
                Log.d(TAG, "asyncLoad onFailed")
            })
        }
        var canceled = false
        var task1: Future<*>? = null
        syncLoad.setOnClickListener {
            canceled = false
            task1 = threadPool.submit {
                Log.d(TAG, "syncLoad1:${loader.syncLoad(URL_BIG, { canceled }) { total, current, speed ->
                    Log.d(TAG, "syncLoad1 onProgress: total:$total    current:$current    speed:$speed")
                }
                }")
            }
        }
        cancel.setOnClickListener {
            Log.d(TAG, "cancel")
            loader.cancel(URL_BIG)
        }
        clear.setOnClickListener {
            loader.clearCache()
        }
        check.setOnClickListener {
            Log.d(TAG, "check:${loader.checkDownload(URL_BIG)}")
        }
    }
}

data class Params(val path: String, val width: Int, val height: Int, val centerCrop: Boolean)

private fun File.decodeBitmap(params: Params): Bitmap? {
    val centerCrop = params.centerCrop && params.height > 0 && params.width > 0
    if (params.height <= 0 && params.width <= 0) {
        Log.e(TAG, "load bitmap without compress :$absolutePath")
    }
    val opts = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(absolutePath, opts)
    val (height, width) = listOf(params.height, params.width).map { return@map if (it <= 0) Int.MAX_VALUE else it }
    val (out, dst) = listOf(opts.outHeight to height, opts.outWidth to width)
            .run { return@run if (centerCrop) minBy { it.first.toFloat() / it.second } else maxBy { it.first.toFloat() / it.second } }!!

    Log.d(TAG,"width:$width")
    Log.d(TAG,"height:$height")
    Log.d(TAG,"outWidth:${opts.outWidth}")
    Log.d(TAG,"outHeight:${opts.outHeight}")
    opts.inSampleSize = out / dst
    opts.inDensity = out
    opts.inTargetDensity = dst
    opts.inScaled = true
    opts.inJustDecodeBounds = false

    return if (!centerCrop) BitmapFactory.decodeFile(absolutePath, opts) else
        try {
            val inSampleSize = out.toFloat() / dst
            val bitmap = BitmapRegionDecoder.newInstance(absolutePath, !canWrite()).decodeRegion(Rect(
                    (opts.outWidth / 2 - width * inSampleSize / 2).toInt(),
                    (opts.outHeight / 2 - height * inSampleSize / 2).toInt(),
                    (opts.outWidth / 2 + width * inSampleSize / 2).toInt(),
                    (opts.outHeight / 2 + height * inSampleSize / 2).toInt()), opts) ?: return null
            Log.d(TAG,"raw height:${bitmap.height}")
            Log.d(TAG,"raw width:${bitmap.width}")
            Bitmap.createScaledBitmap(bitmap, height, width, true).also {
                Log.d(TAG,"bitmap.height:${it.height}")
                Log.d(TAG,"bitmap.width:${it.width}")
            }
        } catch (e: IOException) {
            null
        }
}
